internal/utils/merge.go

```diff
@@ -0,0 +1,57 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package utils
+
+import (
+	"encoding/json"
+	"fmt"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	egv1a1 "github.com/envoyproxy/gateway/api/v1alpha1"
+)
+
+func Merge[T client.Object](original T, patch T, mergeType egv1a1.MergeType) (T, error) {
```

Comment:
non blocking comment: would be good to dedup this and https://github.com/envoyproxy/gateway/blob/645b65fa16c6ed924bcb5c9abbdcd4a60e64e787/api/v1alpha1/kubernetes_helpers.go#L177 in a follow up

Reply:
https://github.com/envoyproxy/gateway/issues/5849